server/src/main/java/org/elasticsearch/index/mapper/NestedFieldValueFetcher.java

```diff
@@ -0,0 +1,128 @@
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+public class NestedFieldValueFetcher implements ValueFetcher {
+
+    private final String nestedFieldPath;
+    private final FieldFetcher nestedFieldFetcher;
+
+    // the name of the nested field without the full path, i.e. in foo.bar.baz it would be baz
+    private final String nestedFieldName;
+    private final String[] nestedPathParts;
+
+    public NestedFieldValueFetcher(String nestedField, FieldFetcher nestedFieldFetcher) {
+        assert nestedField != null && nestedField.isEmpty() == false;
+        this.nestedFieldPath = nestedField;
+        this.nestedFieldFetcher = nestedFieldFetcher;
+        this.nestedPathParts = nestedFieldPath.split("\\.");
+        this.nestedFieldName = nestedPathParts[nestedPathParts.length - 1];
+    }
+
+    @Override
+    public List<Object> fetchValues(SourceLookup lookup, Set<String> ignoreFields) throws IOException {
+        List<Object> nestedEntriesToReturn = new ArrayList<>();
+        Map<String, Object> filteredSource = new HashMap<>();
+        Map<String, Object> stub = createSourceMapStub(filteredSource);
+        List<?> nestedValues = XContentMapValues.extractNestedValue(nestedFieldPath, (lookup.loadSourceIfNeeded()));
+        if (nestedValues == null) {
+            return Collections.emptyList();
+        }
+        List<Map> nestedList = unpack(nestedValues);
+        for (Object entry : nestedList) {
+            // add this one entry only to the stub and use this as source lookup
+            stub.put(nestedFieldName, entry);
+            SourceLookup nestedSourceLookup = new SourceLookup();
+            nestedSourceLookup.setSource(filteredSource);
+
+            Map<String, DocumentField> fetchResult = nestedFieldFetcher.fetch(nestedSourceLookup, ignoreFields);
+
+            Map<String, Object> nestedEntry = new HashMap<>();
+            for (DocumentField field : fetchResult.values()) {
+                List<Object> fetchValues = field.getValues();
+                if (fetchValues.isEmpty() == false) {
+                    String keyInNestedMap = field.getName().substring(nestedFieldPath.length() + 1);
+                    nestedEntry.put(keyInNestedMap, fetchValues);
+                }
+            }
+            if (nestedEntry.isEmpty() == false) {
+                nestedEntriesToReturn.add(nestedEntry);
+            }
+        }
+        return nestedEntriesToReturn;
+    }
+
+    /**
+     * Go through the input list and unpack every entry that is itself a list.
+     * This can happen when e.g. "obj" and "obj2" are both nested and we have something like
+     * "obj" : [
+     *    {
+     *      "obj2 : { ...}
+     *    },
+     *    {
+     *      "obj2 : [ { ... } ]
+     *    }
+     * ]
+     * in the source and get "obj.obj2"
+     */
+    private List<Map> unpack(List<?> nestedValues) {
```

Comment:
I had trouble understanding this 'unpacking' logic, could you explain it a bit more? I also tried removing it and noticed all the tests still passed.

Reply:
Interesting, it was necessary at some point but maybe some recent changes take care of it. I'm pretty sure I have a test for the problematic input document format, will check and see if that indeed passes. I removed this code for now.

Reply:
I did some digging and think I know why I added this but don't need it anymore. For documents like:
```
{
  "user" : [
    {
      "name" : "John Brown",
      "address" : {  <--------- note this would be a Map in the source map
        "city" : "Berlin",
        "zip" : 12345
      }
    },
    {
      "name" : "Alice White",
      "address" : [{ <--------- note this would be a List in the source map
        "city" : "Toronto",
        "zip" : 67890
      },{
        "city" : "Ottawa",
        "zip" : 78901
      }]
    }
  ]
}
```
where both "user" and "user.address" are mapped as `nested`, I first though for requesting "user.adress" we'd like to return a flat list for "user.address.*" like this:
```
"user.address" : [ { "city" : "Berlin", "zip" : 12345 }, { "city" : "Toronto", "zip" : 67890 },{ "city" : "Ottawa", "zip" : 78901 } ]
```
and for that his logic was needed. 
In the current state of the PR we would return a structured "user" array here as well:
```
"user" : [ { address" : [ { "city" : "Berlin", "zip" : 12345 } ] }, { address : [ { "city" : "Toronto", "zip" : 67890 },{ "city" : "Ottawa", "zip" : 78901 } ] }]"
```
and for this the upacking step isn't needed anymore. I removed it.